### PR TITLE
feat(sprint-2): integrate Red Team P2/P3 deliverables into main

### PR DESCRIPTION
**Pull request by this-Demir**
### Description
This pull request merges the `red/sprint2` branch into `main`. It completes the Red Team deliverables for P2 and P3 of the SPM system, delivering core business processes, robust API integration, and enhanced testing infrastructure.

### Key Implementations
- **P2/P3 Business Features:** Completed endpoints and service layers for Student uploads, Deliverables, Sprint mappings, Rubrics, and Coordinator assignments.
---

### Test Attribution Summary
- **Blue Team (Pre-existing)**
- **Red Team (New P2/P3)**

### Backend Execution (Maven / Spring Boot)
- **Total Java Tests Run:** 341
- **Pass Rate:** 100% (341 / 341 passing, 0 Failures, 0 Errors)
- **Coverage Insights:** Includes 233+ Service Layer tests bounding edge-cases (`GroupService`, `AdvisorService`, `InvitationService`) and security validations.

### API Validation & E2E Testing (Postman)
- **Total Automated API Requests:** 57
- **Success Rate:** 98.2% (56 / 57 passing)
  - `Coordinator QA Tests`: 22 / 22 Passed
  - `Sprints & Publish`: 17 / 17 Passed
  - `Auth and RBAC Tests`: 17 / 18 Passed *(1 failure is a test mock mapping issue for OAuth)*
- **Manual GitHub OAuth Testing:** Real OAuth flow successfully verified manually against 2 specific datasets (100% success).
- **Startup:** Application wakes up, links to MySQL DB, and remains stable when fielding validation exceptions.

---

## Checklist
- [x] Compilation & backend build is clean
- [x] Java Unit/Integration Tests pass
- [x] API functionality verified (Newman / Postman)
- [x] OpenAPI specification synced and resolved
- [x] Linter/Formatting applied correctly

[TEST_RESULTS_REPORT.md](https://github.com/user-attachments/files/26662977/TEST_RESULTS_REPORT.md)

### CODE REVIEW

File: backend/src/main/java/com/senior/spm/controller/AdvisorController.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+     * @param groupId UUID of the group
+     * @return 200 with {@link AdvisorRequestResponse} containing requestId and status=CANCELLED
+     */
+    @DeleteMapping("/api/groups/{groupId}/advisor-request")
+    public ResponseEntity<AdvisorRequestResponse> cancelAdvisorRequest(
+            @PathVariable UUID groupId) {
+        UUID requesterUUID = extractPrincipalUUID();
+        AdvisorRequestResponse response = advisorService.cancelAdvisorRequest(groupId, requesterUUID);
+        return ResponseEntity.ok(response);
+    }
```
Comment (Positronium09) [inline backend/src/main/java/com/senior/spm/controller/AdvisorController.java:136]: `/api/groups` should be inside of GroupController not AdvisorController

File: backend/src/main/java/com/senior/spm/controller/AdvisorController.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+ *       {@code /api/advisor/**} path.</li>
+ * </ul>
+ *
+ * <p>Principal extraction: both student and staff JWTs set the principal to the internal UUID
+ * string (via {@code JwtAuthenticationFilter} → {@code claims.get("id", String.class)}).
+ */
+@RestController
+@RequiredArgsConstructor
+@Validated
+public class AdvisorController {
```
Comment (Positronium09) [inline backend/src/main/java/com/senior/spm/controller/AdvisorController.java:49]: RequestMapping("/api/advisor") will be a good change since this controller should only be controlling the endpoints with `/api/advisor/**`

File: backend/src/main/java/com/senior/spm/controller/SanitizationController.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+ * routes require the Coordinator role).
+ *
+ * <p>The sanitization job also runs automatically via {@link SanitizationService#runSanitizationIfWindowClosed()}
+ * (every 5 minutes). This endpoint gives coordinators an on-demand trigger — useful for
+ * early sanitization ({@code force: true}) or immediate execution after the window closes.
+ */
+@RestController
+@RequestMapping("/api/coordinator")
+@RequiredArgsConstructor
+public class SanitizationController {
```
Comment (Positronium09) [inline backend/src/main/java/com/senior/spm/controller/SanitizationController.java:29]: endpoint needs to be inside the coordinator controller not here

File: backend/src/main/java/com/senior/spm/entity/SystemConfig.java
Comment (Positronium09) [inline backend/src/main/java/com/senior/spm/entity/SystemConfig.java:1]: missing endline on the end of the file

File: backend/src/main/java/com/senior/spm/config/GlobalExceptionHandler.java [hunk 4]
```diff
@@ -35,15 +50,137 @@ public ResponseEntity<ErrorMessage> handleAccessDenied(AccessDeniedException ex)
         return ResponseEntity.status(HttpStatus.FORBIDDEN).body(new ErrorMessage(message));
     }
 
-    @ExceptionHandler(RepositoryException.class)
-    public ResponseEntity<ErrorMessage> handleRepositoryException(RepositoryException ex) {
-        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
-                .body(new ErrorMessage("An unexpected error occurred: " + ex.getMessage()));
-    }
-
```
Comment (Positronium09) [inline backend/src/main/java/com/senior/spm/config/GlobalExceptionHandler.java:43]: removed the RepositoryException handler why?

File: backend/src/main/java/com/senior/spm/service/GitHubValidationService.java
Comment (Positronium09) [inline backend/src/main/java/com/senior/spm/service/GitHubValidationService.java:1]: in main branch theres already a github service add the missing methods to it

File: backend/src/main/resources/application.properties.example [hunk 5]
```diff
@@ -7,6 +7,16 @@ spring.jpa.defer-datasource-initialization=true
...
 jwt.secret=
 jwt.expiration=604800000
+
+# AES-256-GCM key for encrypting GitHub PATs and JIRA API tokens at rest (NFR-7).
+# Must be a Base64-encoded value that decodes to exactly 32 bytes (256 bits).
+# Generate a key with: openssl rand -base64 32
+encryption.key=
 encryption.secret=
-GITHUB_CLIENT_ID=
-GITHUB_CLIENT_SECRET=
```
Comment (Positronium09) [inline backend/src/main/resources/application.properties.example:12]: these keys need to be present in the application.properties

File: backend/src/main/resources/data.sql
Comment (Positronium09) [inline backend/src/main/resources/data.sql:1]: removal of newly added example data from data.sql add them from the main branch

**Review by Positronium09 — CHANGES_REQUESTED** [review]
A new directory `dto` containing new dtos is added but the previous ones are left out either move everything to `dto/request/` and `dto/response/` or move every class in `dto/` to the `request/` and `response/` that already exist in the main branch

**Comment by this-Demir** [comment]
# PR #125 Review Response

Thanks for the review! Here's what we've addressed and our reasoning on the others:

## Fixed

- **DTO package cleanup** — moved all classes from `controller/dto/` into `controller/request/` and
  `controller/response/` to align with the existing Blue Team conventions.
- **RepositoryException handler** — restored the missing `@ExceptionHandler(RepositoryException.class)`
  in `GlobalExceptionHandler`. It was accidentally dropped during the merge.
- **application.properties.example** — re-added `GITHUB_CLIENT_ID` and `GITHUB_CLIENT_SECRET` keys that
  were missing from this branch.
- **data.sql** — restored the 3 `password_reset_token` seed rows alongside the new `system_config`
  inserts. Both teams' seed data now coexist.
- **SystemConfig.java** — added missing trailing newline.

---

## Not Changed (with reasoning)

- **`/api/groups/{groupId}/advisor-request` in `AdvisorController`** — these endpoints are part of the P3
  advisor request lifecycle, not generic group management. They belong in `AdvisorController` by
  design. Moving them to `GroupController` would mix P2 group CRUD with P3 advisor flow concerns.

- **Class-level `@RequestMapping("/api/advisor")` on `AdvisorController`** — the controller intentionally
  serves three path prefixes: `/api/advisors` (browse), `/api/groups/{groupId}/advisor-request` (student
  flow), and `/api/advisor/requests/**` (professor flow). Adding a class-level prefix would break the
  first two paths.

- **`SanitizationController` as a separate class** — it already maps to `POST /api/coordinator/sanitize`
  and is protected by the coordinator role via `SecurityConfig`. Merging it into `CoordinatorController`
  is a style preference — keeping it separate follows single responsibility principle and keeps
  `CoordinatorController` from growing further.

- **`GitHubValidationService` vs `GithubService`** — these are two completely different services.
  `GithubService` handles OAuth2 login (code exchange, user profile fetch). `GitHubValidationService`
  handles P2 tool binding — it validates a GitHub PAT against a specific org and checks for `repo`
  scope. Merging them would couple unrelated concerns.

**Comment by Positronium09** [comment]
we dont have a seperate RubricController yes? add them to the GroupController the belong there or make a big Controller class that has all the endpoints since you dont care about what it starts with (namely `/api/<endpoint_group>`

**Comment by Positronium09** [comment]
Since both services are about github its best to keep them together instead of opening new classes

**Comment by this-Demir** [comment]
The remaining 2 suggestions were discussed with the Blue Team lead @InextinguishableFire, as they are design preferences with no impact on production code, they were deferred to the next sprint to save time.

**Comment by Positronium09** [comment]
you might as well have put all the classes in one big java file since its a _design preference_ next time
